src/devices/Dhtxx/DhtSensor.cs

```diff
@@ -75,116 +73,174 @@ public bool TryGetTemperature(out Temperature temperature)
         {
-            var ret = ReadData();
-            relativeHumidity = Humidity;
-            return ret;
+            _protocol = CommunicationProtocol.OneWire;
+            _controller = new GpioController(pinNumberingScheme);
+            _pin = pin;
+            _dhtType = dhtType;
+
+            _controller.OpenPin(_pin);
+            // keep data line HIGH
+            _controller.SetPinMode(_pin, PinMode.Output);
+            _controller.Write(_pin, PinValue.High);
+            // delay 1s to make sure DHT stable
+            DelayHelper.DelayMilliseconds(1000, true);
         }
 
         /// <summary>
-        /// Get the temperature and the relative humidity in the air
+        /// Create a DHT sensor through I2C (Only DHT12)
         /// </summary>
-        /// <param name="temperature">The temperature</param>
-        /// <param name="relativeHumidity">The percentage of relative humidity in the air</param>
-        /// <returns>Returns <c>true</c> if the read is successful</returns>
-        /// <remarks>
-        /// If last read was not successfull, it returns double.MaxValue
-        /// </remarks>
-        public bool TryGetTemperatureAndHumidity(out Temperature temperature, out double relativeHumidity)
+        /// <param name="sensor">I2C Device, like UnixI2cDevice or Windows10I2cDevice</param>
+        public DhtSensor(I2cDevice sensor)
         {
-            var ret = ReadData();
-            temperature = Temperature;
-            relativeHumidity = Humidity;
-            return ret;
+            _protocol = CommunicationProtocol.I2C;
+            _sensor = sensor;
+            _dhtType = DhtType.DHT12;
         }
 
         /// <summary>
-        /// Create a DHT sensor
+        /// Start a reading
         /// </summary>
-        /// <param name="pin">The pin number (GPIO number)</param>
-        /// <param name="dhtType">The DHT Type, either Dht11 or Dht22</param>
-        public DHTSensor(int pin, DhtType dhtType)
+        /// <returns>
+        /// <c>true</c> if read is successfull, otherwise <c>false</c>.
+        /// </returns>
+        private void ReadData()
         {
-            this._pin = pin;
-            this._dhtType = dhtType;
-            _controller.OpenPin(pin);
+            // The time of two measurements should be more than 1s.
+            if (Environment.TickCount - _lastMeasurment < 1000)
+            {
+                return;
+            }
+
+            if (_protocol == CommunicationProtocol.OneWire)
+            {
+                ReadThroughOneWire();
+            }
+            else
+            {
+                ReadThroughI2c();
+            }
         }
 
         /// <summary>
-        /// Start a reading
+        /// Read through One-Wire
         /// </summary>
         /// <returns>
         /// <c>true</c> if read is successfull, otherwise <c>false</c>.
         /// </returns>
-        public bool ReadData()
+        private bool ReadThroughOneWire()
```

Comment:
would be nice to merge this with: https://github.com/dotnet/iot/pull/289 eventually (not in this PR though)